airflow/models/taskinstance.py

```diff
@@ -1521,7 +1524,25 @@ def _run_raw_task(
         if not test_mode:
             session.add(Log(self.state, self))
             session.merge(self)
-
+            for obj in getattr(self.task, '_outlets', []):
+                self.log.debug("outlet obj %s", obj)
+                if isinstance(obj, Dataset):
+                    dataset = session.query(Dataset).filter(Dataset.uri == obj.uri).first()
+                    if not dataset:
+                        dataset = Dataset(uri=obj.uri, extra=obj.extra)
+                        self.log.debug("adding dataset %s", dataset)
+                        session.add(dataset)
+                        session.flush()
```

Comment:
How can this be hit? Shouldn't it have been added during parse?

Reply:
probably true... maybe in the mapped case?   i suppose we could chop the defensiveness until it's apparent that it is required